README.md
  TurnsTile
  =========

  A mosaic and palette effects plugin for [Avisynth](http://www.avisynth.nl).

  [![Build Status](https://dev.azure.com/robertmartens0491/TurnsTile/_apis/build/status/ItEndsWithTens.TurnsTile?repoName=ItEndsWithTens%2FTurnsTile&branchName=master)](https://dev.azure.com/robertmartens0491/TurnsTile/_build/latest?definitionId=4&repoName=ItEndsWithTens%2FTurnsTile&branchName=master)

  ### Requirements ###
  1. Avisynth 2.6 or greater, or Avisynth+
  2. [**Windows**] Microsoft's Visual C++ 2019 Redistributable - [x64](https://aka.ms/vs/16/release/VC_redist.x64.exe) - [x86](https://aka.ms/vs/16/release/VC_redist.x86.exe)

  http://www.gyroshot.com/turnstile.htm  
  robert.martens@gmail.com  
  [@ItEndsWithTens](https://twitter.com/ItEndsWithTens)

  ### TurnsTile ###

    TurnsTile(clip c, clip "tilesheet", int "tilew", int "tileh", int "res",
              int "mode", string "levels", int "lotile", int "hitile",
              bool "interlaced")

  **c** clip
  - The input clip, which can be RGB32, RGB24, YUY2, or YV12.

  **tilesheet** clip
  - Optional; if supplied, tiles will be pulled from this clip, which must be in  
    the same colorspace as 'c'. Your tilesheet can be a still image or a video.  
    In the latter case, tiles for a given frame of 'c' will be clipped from the  
    corresponding frame of 'tilesheet'.

    The tiles are numbered left to right, then top to bottom. Using the provided  
    tilesheets as examples, with 16x16 pixel tiles in a 256x256 pixel image, the  
    top left tile is number 0, top right is 15, bottom left is 240, and bottom  
    right is 255. Design your own custom images accordingly, with dark tones at  
    the top left fading up to lighter ones at the bottom right (first left to  
    right across a row, then top to bottom one row at a time).

  **tilew, tileh** int, default largest size <= 16x16 that fits your input
  - If your tiles aren't sixteen by sixteen, define custom values here. Each  
    must be a factor of the respective clip dimension.

  **res** int, default 8
  - This acts as the effective bit depth of your output. The range of possible  
    output values is broken up into 2 ** res steps, and each tile index or pixel  
    component is rounded accordingly. This is quite an effective technique for  
    RGB footage, but thanks to the way color is represented in YUV spaces, you  
    won't be able to lower this too much with YUY2 or YV12 input before things  
    begin to look strange.

  **mode** integer, default 0
  - Only works when tilesheet is supplied. This option chooses the component  
    that will serve as the tile index for a given tile. The possible values are  
    as follows:

        0:
          RGB:   Average of red, green, and blue values
          YUY2:  Average of Y1 and Y2 in the current pixel pair
          YV12:  Average of all four Y values in the 2x2 block

        1:
          RGB:   Blue
          YUY2:  Y1
          YV12:  Top left

        2:
          RGB:   Green
          YUY2:  U
          YV12:  Top right

        3:
          RGB:   Red
          YUY2:  Y2
          YV12:  Bottom left

        4:
          RGB32: Alpha
          RGB24: N/A
          YUY2:  V
          YV12:  Bottom right

        5:
          RGB32: N/A
          RGB24: N/A
          YUY2:  N/A
          YV12:  U

        6:
          RGB32: N/A
          RGB24: N/A
          YUY2:  N/A
          YV12:  V

  **levels** string, "pc" or "tv", default "pc"
  - Which range to use when selecting tiles. If you'd like to map TV black and  
    white to the lowest and highest tiles in your tilesheet, respectively, use  
    "tv" instead of the default "pc".

  **lotile, hitile** default 0 and tilecount - 1
  - A quick way to limit the tile selection to a given portion of your  
    tilesheet; if for example you have a sheet with an odd number of tiles, and  
    some spaces are blank, or if you just want to use a smaller range of values  
    without having to rebuild your tilesheet by hand, use these. If you don't  
    use a tilesheet, these will instead control the maximum and minimum  
    component values.

  **interlaced** bool, default false
  - Enable for interlaced input. For those unaware, "interlaced" and "field  
    based" are not the same thing. If a clip is field based, it's more than  
    likely interlaced, but the reverse isn't true, and there's currently no  
    completely fool proof way to auto-detect interlaced input.

  ----

  ### CLUTer ###
    CLUTer(clip c, clip palette, int "paletteframe", bool "interlaced")

  **c** clip
  - No special restrictions, beyond ensuring that this clip's colorspace  
    matches that of your palette. RGB32, RGB24, YUY2, and YV12 are supported.

  **palette** clip
  - Must be the same colorspace as the input clip. Progressive chroma siting is  
    assumed for YV12 palettes; although 'c' will be treated as interlaced if the  
    appropriate option is enabled, the palette itself will always be read as if  
    progressive.

    **CAUTION!** It is very, very easy to dramatically slow down the operation  
    of CLUTer, to such a degree that it may appear to be frozen. The short  
    version: stick with ~100 or fewer unique colors per palette, or you're in  
    for a long wait. If you want to use an image as your palette, run it through  
    TurnsTile first, with big tiles and/or a lowered 'res'.

  **paletteframe** int, default 0
  - Only one frame is used from any clip you pass in as your palette, so if you  
    don't want to use the colors of frame 0, set paletteframe accordingly.

  **interlaced** bool, default false
  - As explained above, the terms "field based" and "interlaced" are not  
    necessarily synonymous. Reading the proper luma values for a given chroma  
    sample requires knowing the nature of the input clip, and a user-defined  
    parameter is the most reliable way to achieve that.

  ----

  ### Extras ###

  Included in the 'extras' directory is a set of tilesheets meant to serve as a  
  jumping off point for your own experimentation, along with the AviSynth script  
  used to generate them. It needs [Gavino's GScript](http://forum.doom9.org/showthread.php?t=147846) to make it work in Avisynth  
  2.6, but there are no other requirements.

  CGApalette.avs was introduced along with CLUTer() in version 0.3.0, and as  
  stated in the comments is meant to be loaded by AVISource as a sample palette  
  for that function. Open up the script to find more details.

  Lastly, also in the extras folder is a copy of the script version of TurnsTile  
  0.1.0. Use in Aviysnth 2.6 requires GScript, along with [GRunT](http://forum.doom9.org/showthread.php?t=139337), also by Gavino,  
  and doesn't run very quickly, to say the least. That barely measurable speed  
  was the motivation to develop this plugin, and the script is only included  
  here for educational purposes.


CHANGELOG.md
# Changelog
Changes listed according to the [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) standard. This project attempts to use [Semantic Versioning](https://semver.org/spec/v2.0.0.html), to the best of its author's ability.

## [1.0.0] 2020-07-16
### Added
- Add 64-bit builds
- Add Avisynth+ Linux and macOS support

### Changed
- Build with Avisynth+ header

### Removed
- Remove Avisynth 2.5 support

## [0.3.2] 2013-04-10
### Fixed
- Fix tilesheet processing for interlaced clips

## [0.3.1] 2011-09-21
### Changed
- Recompile to remove SSE/SSE2 requirement and Visual C++ CRT dependency

## [0.3.0] 2011-05-06
### Added
- Add YV12 support
- Add CLUTer palettizing function

### Changed
- Expand lotile/hitile operation to clip-only mode

### Fixed
- Correct braindead inclusion guard mistake in TurnsTile.h
- Fix 'res' option to round values correctly

## [0.2.1] 2011-03-12
### Fixed
- Update auto calc for tile size, now properly accounts for minimum YUY2 width

## [0.2.0] 2011-03-10
### Added
- Add 'levels', 'lotile', 'hitile'
- Add auto calculation of default tile size

### Changed
- Change 'res' behavior to simulate bit depth adjustment

## [0.1.0] 2010-12-24
- Initial release


src/interface.cpp
#include "CLUTer.h"
#include "TurnsTile.h"
#include "TurnsTileTestSource.h"

#include <cmath>
#include <cstring>

#include "interface.h"



AVSValue __cdecl Create_TurnsTile(
  AVSValue args, void* user_data, IScriptEnvironment* env)
{

  PClip clip = args[0][0].AsClip(),
        tilesheet = 0;
  if (args[0].ArraySize() > 1)
    tilesheet = args[0][1].AsClip();
  VideoInfo vi = clip->GetVideoInfo(),
            vi2 = vi;
  if (tilesheet)
    vi2 = tilesheet->GetVideoInfo();


  int dTileW = 16,
      dTileH = 16;

  int clipW = vi.width,
      clipH = vi.height,
      sheetW = 0,
      sheetH = 0;
  if (tilesheet) {
    sheetW = vi2.width;
    sheetH = vi2.height;
  }

  int lumaW, lumaH;

  if (vi.IsYUV() && !vi.IsY8()) {
    lumaW = 1 << vi.GetPlaneWidthSubsampling(PLANAR_U);
    lumaH = 1 << vi.GetPlaneHeightSubsampling(PLANAR_U);
  } else {
    lumaW = 1;
    lumaH = 1;
  }


  // Reading arguments out of order makes me feel icky, but I need this early.
  bool interlaced = args[8].AsBool(false);


  int minTileW = lumaW,
      minTileH = lumaH;
  if (interlaced)
    minTileH *= 2;

  // I've saved most of my error handling for later, but I have to check for a
  // divide by zero that might crop up when reducing dTileW and dTileH. The
  // matching colorspace check is here because I feel like it should be first.
  if (!vi.IsSameColorspace(vi2))
      env->ThrowError("TurnsTile: clip and tilesheet must share a colorspace!");

  const char* const interlacedStr = interlaced ? "interlaced " : "";
  const char* const cspStr =  vi.IsRGB32() ?  "RGB32" :
                              vi.IsRGB24() ?  "RGB24" :
                              vi.IsYUY2() ?   "YUY2" :
                              vi.IsYV12() ?   "YV12" :
                              vi.IsYV24() ?   "YV24" :
                              vi.IsYV16() ?   "YV16" :
                              vi.IsYV411() ?  "YV411" :
                              vi.IsY8() ?     "Y8" :
                              interlaced ?    "" :
                                              "this";

  if (interlaced) {

    if (clipH % minTileH > 0)
      env->ThrowError(
        "TurnsTile: %s clip height must be mod %d when interlaced=true!",
        cspStr, minTileH);

    if (sheetH % minTileH > 0)
      env->ThrowError(
        "TurnsTile: %s tilesheet height must be mod %d when interlaced=true!",
        cspStr, minTileH);

  }

  // Reduce each default tile dimension to the greatest size, less than or equal
  // to its starting value, that's a factor of the clip and tilesheet dimension,
  // and a multiple of the minimum tile size.
  while (clipW % dTileW > 0 || sheetW % dTileW > 0 || dTileW % minTileW > 0)
    --dTileW;
  while (clipH % dTileH > 0 || sheetH % dTileH > 0 || dTileH % minTileH > 0)
    --dTileH;

  // Try to get square tiles, if possible.
  if (dTileW != dTileH && clipW % dTileH == 0 && sheetW % dTileH == 0)
    dTileW = dTileH;
  if (dTileH != dTileW && clipH % dTileW == 0 && sheetH % dTileW == 0)
    dTileH = dTileW;

  int tileW = args[1].AsInt(dTileW),
      tileH = args[2].AsInt(dTileH);

  // Another pair of out of place error checks for another possible divide by
  // zero, this time when calculating tileIdxMax a few lines down.
  if (tileW < minTileW)
    env->ThrowError(
      "TurnsTile: tilew must be at least %d for %s input!",
      minTileW, cspStr);

  if (tileH < minTileH)
    env->ThrowError(
      "TurnsTile: tileh must be at least %d for %s%s input!",
      minTileH, interlacedStr, cspStr);


  int res = args[3].AsInt(8);


  int mode = args[4].AsInt(0);


  const char* levels = env->Invoke("LCase",args[5].AsString("pc")).AsString();


  int loTile = args[6].AsInt(0);

  // Doesn't make a difference here whether interlaced is true or not, since if
  // that's the case we'll later be halving both the tilesheet height and tile
  // height. A / B == (A/2) / (B/2), so tileIdxMax is the same either way.
  int tileIdxMax;
  if (tilesheet)
    tileIdxMax = ((sheetW / tileW) * (sheetH / tileH)) - 1;
  else
    tileIdxMax = 255;

  int hiTile = args[7].AsInt(tileIdxMax);


  int maxTileW = TurnsTile::gcf(clipW, sheetW),
      maxTileH = TurnsTile::gcf(clipH, sheetH);

  // These two errors, unlike the two above and the two below, don't mention
  // anything about interlacing or colorspace since the check is performed based
  // only on the greatest common factor of the clip and tilesheet width/height.
  if (tileW > maxTileW)
    env->ThrowError(
      "TurnsTile: For this input, tilew must not exceed %d!",
      maxTileW);

  if (tileH > maxTileH)
    env->ThrowError(
      "TurnsTile: For this input, tileh must not exceed %d!",
      maxTileH);

  if (tileW % minTileW > 0)
    env->ThrowError(
      "TurnsTile: For %s input, tilew must be a multiple of %d!",
      cspStr, minTileW);

  if (tileH % minTileH > 0)
    env->ThrowError(
      "TurnsTile: For %s%s input, tileh must be a multiple of %d!",
      interlacedStr, cspStr, minTileH);

  if (maxTileW % tileW > 0)
    env->ThrowError(
      "TurnsTile: For this input, tilew must be a factor of %d!",
      maxTileW);

  if (maxTileH % tileH > 0)
    env->ThrowError(
      "TurnsTile: For this input, tileh must be a factor of %d!",
      maxTileH);


  int countChroma = 2;
  if (vi.IsY8())
    countChroma = 0;

  int modeMax;
  if (vi.IsYUV())
    modeMax = (lumaW * lumaH) + countChroma;
  else
    modeMax = vi.BytesFromPixels(1);

  if (mode < 0 || mode > modeMax)
    env->ThrowError(
      "TurnsTile: %s only allows modes 0-%d!",
      cspStr, modeMax);
  if ((vi.IsYV24() || vi.IsY8()) && mode == 0)
    mode = 1;


  if (strcmp(levels, "pc") != 0 && strcmp(levels, "tv") != 0)
    env->ThrowError(
      "TurnsTile: levels must be either \"pc\" or \"tv\"!");


  if (loTile < 0 || loTile > tileIdxMax)
    env->ThrowError(
      "TurnsTile: Valid lotile range is 0-%d!",
      tileIdxMax);

  if (hiTile < 0 || hiTile > tileIdxMax)
    env->ThrowError(
      "TurnsTile: Valid hitile range is 0-%d!",
      tileIdxMax);

  if (loTile > hiTile)
    env->ThrowError(
      "TurnsTile: lotile must not be greater than hitile!");


  if (interlaced) {

    tileH /= 2;
    if (!clip->GetVideoInfo().IsFieldBased())
      clip = env->Invoke("SeparateFields", clip).AsClip();
    if (tilesheet && !tilesheet->GetVideoInfo().IsFieldBased()) {
      tilesheet = env->Invoke("SeparateFields", tilesheet).AsClip();
      vi2 = tilesheet->GetVideoInfo();
    }

  }

  PClip finalClip = new TurnsTile(  clip,
                                    tilesheet,
                                    vi2,
                                    tileW,
                                    tileH,
                                    res,
                                    mode,
                                    levels,
                                    loTile,
                                    hiTile,
                                    env);

  if (interlaced && finalClip->GetVideoInfo().IsFieldBased())
    return env->Invoke("Weave", finalClip);
  else
    return finalClip;

}



AVSValue __cdecl Create_CLUTer(
  AVSValue args, void* user_data, IScriptEnvironment* env)
{

  PClip clip = args[0].AsClip(),
        palette = args[1].AsClip();
  VideoInfo vi = clip->GetVideoInfo();


  int paletteFrame = args[2].AsInt(0);


  bool interlaced = args[3].AsBool(false);


  if (!vi.IsSameColorspace(args[1].AsClip()->GetVideoInfo()))
    env->ThrowError("CLUTer: clip and palette must share a colorspace!");


  if (interlaced) {

    const char* const cspStr =  vi.IsRGB32() ?  "RGB32" :
                                vi.IsRGB24() ?  "RGB24" :
                                vi.IsYUY2() ?   "YUY2" :
                                vi.IsYV12() ?   "YV12" :
                                vi.IsYV24() ?   "YV24" :
                                vi.IsYV16() ?   "YV16" :
                                vi.IsYV411() ?  "YV411" :
                                vi.IsY8() ?     "Y8" :
                                                "";

    int minClipH;
    if (vi.IsYV12())
      minClipH = 4;
    else
      minClipH = 2;

    if (vi.height % minClipH != 0)
      env->ThrowError(
        "CLUTer: %s clip height must be mod %d when interlaced=true!",
        cspStr, minClipH);

    if (!vi.IsFieldBased())
      clip = env->Invoke("SeparateFields", clip).AsClip();

  }


  PClip finalClip = new CLUTer(  clip,
                                 palette,
                                 paletteFrame,
                                 interlaced,
                                 env);

  if (interlaced && finalClip->GetVideoInfo().IsFieldBased())
    return env->Invoke("Weave", finalClip);
  else
    return finalClip;

}



AVSValue __cdecl Create_TurnsTileTestSource(
  AVSValue args, void* user_data, IScriptEnvironment* env)
{

  std::string filename = args[0].AsString(""),
              pixel_type = args[1].AsString("RGB32");

  return new TurnsTileTestSource(filename, pixel_type, env);

}



const AVS_Linkage* AVS_linkage = 0;
extern "C" __declspec(dllexport) const char* __stdcall AvisynthPluginInit3(
  IScriptEnvironment* env, const AVS_Linkage* const vectors)
{

  AVS_linkage = vectors;

  env->AddFunction("CLUTer", "cc[paletteframe]i[interlaced]b",
                             Create_CLUTer, 0);

  env->AddFunction("TurnsTile", "c+[tileW]i[tileH]i[res]i[mode]i[levels]s"
                                "[lotile]i[hitile]i[interlaced]b",
                                Create_TurnsTile, 0);

  env->AddFunction("TurnsTileTestSource", "s[pixel_type]s",
                                          Create_TurnsTileTestSource, 0);

  return "`TurnsTile' - Mosaic and palette effects";

}
